modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java

```diff
@@ -677,7 +677,54 @@ public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
         return new NumberFieldMapper.NumericSyntheticFieldLoader(name(), simpleName()) {
             @Override
             protected void loadNextValue(XContentBuilder b, long value) throws IOException {
-                b.value(value / scalingFactor);
+                b.value(convert(value));
+            }
+
+            /**
+             * Convert the scaled value back into it's {@code double} representation,
+             * attempting to undo {@code Math.round(value * scalingFactor)}. It's
+             * important that "round tripping" a value into the index, back out, and
+             * then back in yields the same index. That's important because we use
+             * the synthetic source produced by these this for reindex.
+             * <p>
+             * The tricky thing about undoing {@link Math#round} is that it
+             * "saturates" to {@link Long#MAX_VALUE} when the {@code double} that
+             * it's given is too large to fit into a {@code long}. And
+             * {@link Long#MIN_VALUE} for {@code double}s that are too small. But
+             * {@code Long.MAX_VALUE / scalingFactor} doesn't always yield a value
+             * that would saturate. In other words, sometimes:
+             * <pre>{@code
+             *   long scaled1 = Math.round(BIG * scalingFactor);
+             *   assert scaled1 == Long.MAX_VALUE;
+             *   double decoded = scaled1 / scalingFactor;
+             *   long scaled2 = Math.round(decoded * scalingFactor);
+             *   assert scaled2 != Long.MAX_VALUE;
+             * }</pre>
+             * <p>
+             * We work around this by detecting such cases and artificially bumping them
+             * up by a single digit in the last place, forcing them to always saturate
+             * the {@link Math#round} call.
+             */
+            private double convert(long scaledValue) {
+                if (scaledValue == Long.MAX_VALUE) {
+                    double max = Long.MAX_VALUE / scalingFactor;
+                    if (Math.round(max * scalingFactor) != Long.MAX_VALUE) {
+                        double v = max + Math.ulp(max);
+                        assert Math.round(v * scalingFactor) == Long.MAX_VALUE;
+                        return v;
+                    }
+                    return max;
+                }
+                if (scaledValue == Long.MIN_VALUE) {
+                    double min = Long.MIN_VALUE / scalingFactor;
+                    if (Math.round(min * scalingFactor) != Long.MIN_VALUE) {
+                        double v = min - Math.ulp(min);
+                        assert Math.round(v * scalingFactor) == Long.MIN_VALUE;
+                        return v;
+                    }
+                    return min;
+                }
+                return scaledValue / scalingFactor;
```

Comment:
Given this logic is fairly gnarly, is it worth extracting it into a helper and then throwing some heavy-duty testing on it? Even something quick-check like, to test boundary conditions and then a solid random selection of points.

Reply:
We get that from the round trip tests but I could certainly see wanting it to be tested more directly. I'll yank it around a bit.